node/src/components/contract_runtime/operations.rs

```diff
@@ -99,3 +107,44 @@ pub(super) async fn execute(
     trace!(?result, "execute result");
     result
 }
+
+/// Gets the deploy(s) of the given finalized block from storage.
+pub(super) async fn get_deploys_and_finalize_block<REv>(
+    effect_builder: EffectBuilder<REv>,
+    finalized_block: FinalizedBlock,
+) -> Option<Event>
+where
+    REv: From<ControlAnnouncement> + From<StorageRequest>,
+{
+    // Get the deploy hashes for the finalized block.
+    let deploy_hashes = finalized_block
+        .deploys_and_transfers_iter()
+        .copied()
+        .collect::<SmallVec<_>>();
+
+    let era_id = finalized_block.era_id();
+    let height = finalized_block.height();
+
+    // Get all deploys in order they appear in the finalized block.
+    let mut deploys: VecDeque<Deploy> = VecDeque::with_capacity(deploy_hashes.len());
+    for maybe_deploy in effect_builder.get_deploys_from_storage(deploy_hashes).await {
+        if let Some(deploy) = maybe_deploy {
+            deploys.push_back(deploy)
+        } else {
+            fatal!(
+                effect_builder,
+                "deploy for block in era={} and height={} is expected to exist in the storage",
+                era_id,
+                height
+            )
```

Comment:
~Can we really expect that we have all those deploys in storage?~
~Consensus will currently synchronize the protocol state of the past six eras, and will send those "finalized" events for all their blocks.~

Ah, the reason is that we synchronize the global state for the _oldest_ block we downloaded, and then run all the deploys, so we _will_ have them, right? (Just another invariant we need to make sure we never break.)